under latest MinGW, cannot link with C code using stdout

I'm having an issue under MinGW and I'm not sure if it's a MinGW issue or a rustc issue. On the one hand the bug crops up after an update to two MinGW packages, but on the other hand I cannot reproduce with only C, and I don't know if rustc is using MinGW in some undocumented unsupported way.

say_hi.c:
```c
#include <stdio.h>
void say_hi(void) {
    fprintf(stdout, "hi!\n");
}
```
c_main.c
```c
void say_hi(void);
int main(void) {
    say_hi();
    return 0;
}
```
r_main.rs:
```rust
extern "C" {
    fn say_hi();
}
fn main() {
    unsafe {
        say_hi();
    }
}
```

Under the latest MinGW, specifically with packages

* `mingw-w64-x86_64-crt-git-6.0.0.5066.61efe559-1` and
* `mingw-w64-x86_64-headers-git-6.0.0.5066.61efe559-1`,

using only gcc like this works:
```
$ gcc -c say_hi.c
$ ar cr libsay_hi.a say_hi.o
$ gcc c_main.c -L. -lsay_hi -o c_main.exe && ./c_main.exe
hi!
```
But it doesn't work with rustc:
```
$ gcc -c say_hi.c
$ ar cr libsay_hi.a say_hi.o
$ rustc r_main.rs -L. -lsay_hi -o r_main.exe && ./r_main.exe
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\crt2.o" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "r_main.r_main0.rust-cgu.o" "-o" "r_main.exe" "r_main.crate.allocator.rust-cgu.o" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "." "-L" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-l" "say_hi" "-Wl,-Bstatic" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-60d4a252bdda9250.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc_system-a05404e5e059417a.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librand-81cbf6f1d5a19206.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-0b63e20460046692.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-f824a78e715239af.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd_unicode-2abaea46f194a6e1.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-7daa32005d584d8a.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-49ff25caaa6e945f.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-f2bd2d524c6ace24.rlib" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-c6139fac5e899293.rlib" "-Wl,-Bdynamic" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,-Bstatic" "-l" "gcc_eh" "-l" "pthread" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\Users\\Trevor\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: ./libsay_hi.a(say_hi.o):say_hi.c:(.text+0x10): undefined reference to `__imp___acrt_iob_func'
          

error: aborting due to previous error
```

If I downgrade those two packages to

* `mingw-w64-x86_64-crt-git-5.0.0.5002.34a7c1c0-1` and
* `mingw-w64-x86_64-headers-git-5.0.0.5002.34a7c1c0-1`,

it works with rustc too:
```
$ gcc -c say_hi.c
$ ar cr libsay_hi.a say_hi.o
$ rustc r_main.rs -L. -lsay_hi -o r_main.exe && ./r_main.exe
hi!
```

One last thing, if I compile the library using the old packages, and then upgrade the packages, both `gcc c_main.c ...` and `rustc r_main.rs ...` work. If I compile the library using the new packages, and then downgrade the packages, both `gcc c_main.c ...` and `rustc r_main.rs ...` fail.

I'm not sure who is responsible for Windows GNU, @retep998?

I assume libs from https://github.com/rust-lang/rust/issues/47359 are necessary for people who don't have GCC in their PATH. However they are an issue for people with GCC/MinGW versions different from Rust ones.
They are built in specific environment and the fact they work in (some) other environments is just a coincidence that shouldn't be relied upon. Yet Windows GNU target prioritises them over host libraries and relies on host headers causing conflicts like `undefined reference to `__imp___acrt_iob_func'` from this issue.

It can be solved by:
- using host libraries over bundled if they exists
- bundling headers (66 MiB) and using them instead of host

@retep998 Well, it seems that it still get undefined reference...
```
  = note: C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../
x86_64-w64-mingw32/bin/ld.exe: C:\Users\Administrator\.rustup\toolchains\stable-
x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-gnu\lib\crt2.o:crtexe.c:(.r
data$.refptr.__onexitbegin[.refptr.__onexitbegin]+0x0): undefined reference to `
__onexitbegin'
          C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../
x86_64-w64-mingw32/bin/ld.exe: C:\Users\Administrator\.rustup\toolchains\stable-
x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-gnu\lib\crt2.o:crtexe.c:(.r
data$.refptr.__onexitend[.refptr.__onexitend]+0x0): undefined reference to `__on
exitend'
          collect2.exe: error: ld returned 1 exit status
```
I have install VC++ build tools, but why shouldn't I directly use msvc toolchain to compile it instead of cross compile?:joy:

>Easiest workaround is to just copy crt2.o and dllcrt2.o from C:\msys64\mingw64\x86_64-w64-mingw32\lib to ~\.rustup\toolchains\stable-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib every time you update.

I upgraded to `1.35.0` today and this did _not_ solve the issue. I receive the following error when linking the code in the original post:

```
$ rustc r_main.rs -L. -lsay_hi -o r_main.exe && ./r_main.exe
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\crt2.o" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "r_main.r_main.7rcbfp3g-cgu.0.rcgu.o" "r_main.r_main.7rcbfp3g-cgu.1.rcgu.o" "r_main.r_main.7rcbfp3g-cgu.2.rcgu.o" "r_main.r_main.7rcbfp3g-cgu.3.rcgu.o" "r_main.r_main.7rcbfp3g-cgu.4.rcgu.o" "-o" "r_main.exe" "r_main.4rwig1jpuebwbe6m.rcgu.o" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "." "-L" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-lsay_hi" "-Wl,--start-group" "-Wl,-Bstatic" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-c954a2d44059a032.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-44ed309e7c0a24e8.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libbacktrace_sys-5562a4bb4d42cef6.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_demangle-f8fcbc8dafb5c415.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libhashbrown-a84a86328accb241.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_std_workspace_alloc-ec9721f567a35b2f.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-06daa6384d17722d.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-acd44a96ed61810f.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-57bfdb57672b491d.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librustc_std_workspace_core-4843b04b2fd1dcfe.rlib" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-51f4f445efc13497.rlib" "-Wl,--end-group" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-eacb47c3ec0f8830.rlib" "-Wl,-Bdynamic" "-ladvapi32" "-lws2_32" "-luserenv" "-Wl,-Bstatic" "-lgcc_eh" "-lpthread" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\Users\\William\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: C:\Users\William\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib\crt2.o: In function `pre_c_init':
          C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:138: undefined reference to `__p__fmode'
          C:\Users\William\.rustup\toolchains\nightly-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\lib\crt2.o: In function `__tmainCRTStartup':
          C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:291: undefined reference to `__p__acmdln'
          ./libsay_hi.a(say_hi.o):say_hi.c:(.text+0x10): undefined reference to `__imp___acrt_iob_func'


error: aborting due to previous error
```

I copied the libraries just fine it seems:
```
William@William-THINK MINGW64 ~/src/rust-crt
$ diff -q /c/msys64/mingw64/x86_64-w64-mingw32/lib/crt2.o /c/Users/William/.rustup/toolchains/nightly-x86_64-pc-windows-gnu/lib/rustlib/x86_64-pc-windows-gnu/lib/crt2.o

William@William-THINK MINGW64 ~/src/rust-crt
$ diff -q /c/msys64/mingw64/x86_64-w64-mingw32/lib/dllcrt2.o /c/Users/William/.rustup/toolchains/nightly-x86_64-pc-windows-gnu/lib/rustlib/x86_64-pc-windows-gnu/lib/dllcrt2.o
```

The C link demonstrated in the original post works fine:
```
William@William-THINK MINGW64 ~/src/rust-crt
$ gcc c_main.c -L. -lsay_hi -o c_main.exe && ./c_main.exe
```

GCC version is 7.3.0:
```
William@William-THINK MINGW64 ~/src/rust-crt
$ gcc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/7.3.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-7.3.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,objc,obj-c++,fortran,ada --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-time=yes --enable-libstdcxx-filesystem-ts=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev2, Built by MSYS2 project' --with-bugurl=https://sourceforge.net/projects/msys2 --with-gnu-as --with-gnu-ld
Thread model: posix
gcc version 7.3.0 (Rev2, Built by MSYS2 project)
```

And here is my `~/.cargo/config`:
```
William@William-THINK MINGW64 ~/src/rust-crt
$ cat /c/Users/William/.cargo/config
[target.x86_64-pc-windows-gnu]
linker = "C:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-gcc"
```

Am I missing something obvious to get Rust working again?

>It's still not fixed and this workaround is recommend.
But the workaround _did not_ work. Even after I copied the relevant libraries, I still get the ` undefined reference to `__imp___acrt_iob_func'` error, among others. The information in my previous message is to try to isolate why the workaround didn't work.

Anyways, I'm fully upgrading my msys2 distribution to see if this helps.

### For whoever is getting here with the original problem, the following worked for me: 
#### (`undefined reference to '__imp___acrt_iob_func'`)

## 1. Create workaround.c with the following
###### (thanks @tspiteri for the workaround code)

```c
/* workaround.c */
#define _CRTBLD
#include <stdio.h>

FILE *__cdecl __acrt_iob_func(unsigned index)
{
    return &(__iob_func()[index]);
}

typedef FILE *__cdecl (*_f__acrt_iob_func)(unsigned index);
_f__acrt_iob_func __MINGW_IMP_SYMBOL(__acrt_iob_func) = __acrt_iob_func;
```

## 2. Compile and create the library
```bash
gcc -fPIC -O2 -c workaround.c
ar cr libworkaround_47048.a workaround.o
```

## 3. Move the library to rust's lib folder
```bash
mv libworkaround_47048.a ~/.rustup/toolchains/stable-x86_64-pc-windows-gnu/lib/rustlib
/x86_64-pc-windows-gnu/lib/

rm workaround.c workaround.o
```

## 4. Add the following to `~/.cargo/config`:
```toml
[target.x86_64-pc-windows-gnu]
rustflags=["-Clink-args=-lworkaround_47048"]
```

Enjoy 😄

@daniel-shimon & @tspiteri: this workaround is not compiling for me. It gives the following 2 warnings and 2 errors:
```
workaround.c: In function ‘__acrt_iob_func’:
workaround.c:7:14: warning: implicit declaration of function ‘__iob_func’ [-Wimplicit-function-declaration]
     return &(__iob_func()[index]);
              ^
workaround.c:7:26: error: subscripted value is neither array nor pointer nor vector
     return &(__iob_func()[index]);
                          ^
workaround.c: At top level:
workaround.c:11:1: warning: parameter names (without types) in function declaration
 _f__acrt_iob_func __MINGW_IMP_SYMBOL(__acrt_iob_func) = __acrt_iob_func;
 ^
workaround.c:11:1: error: function ‘__MINGW_IMP_SYMBOL’ is initialized like a variable
```

Updating from gcc 5.4.0 to gcc 7.4.0 did not fix this (on MinGW 64 bit, Windows 7).

> ### For whoever is getting here with the original problem, the following worked for me:
> #### (`undefined reference to '__imp___acrt_iob_func'`)
> ## 1. Create workaround.c with the following
> ###### (thanks @tspiteri for the workaround code)
> ```c
> /* workaround.c */
> #define _CRTBLD
> #include <stdio.h>
> 
> FILE *__cdecl __acrt_iob_func(unsigned index)
> {
>     return &(__iob_func()[index]);
> }
> 
> typedef FILE *__cdecl (*_f__acrt_iob_func)(unsigned index);
> _f__acrt_iob_func __MINGW_IMP_SYMBOL(__acrt_iob_func) = __acrt_iob_func;
> ```
> 
> ## 2. Compile and create the library
> ```shell
> gcc -fPIC -O2 -c workaround.c
> ar cr libworkaround_47048.a workaround.o
> ```
> 
> ## 3. Move the library to rust's lib folder
> ```shell
> mv libworkaround_47048.a ~/.rustup/toolchains/stable-x86_64-pc-windows-gnu/lib/rustlib
> /x86_64-pc-windows-gnu/lib/
> 
> rm workaround.c workaround.o
> ```
> 
> ## 4. Add the following to `~/.cargo/config`:
> ```toml
> [target.x86_64-pc-windows-gnu]
> rustflags=["-Clink-args=-lworkaround_47048"]
> ```
> 
> Enjoy 😄

Hi Daniel @daniel-shimon , I tried your solution, but unfortunately I'm getting this error: 

> = note: ld: cannot find -lworkaround_47048

What am I doing wrong? Any idea of how can I solve this? Thanks in advance.

--- a/src/librustc_codegen_ssa/back/link.rs
+++ b/src/librustc_codegen_ssa/back/link.rs
@@ -968,7 +968,78 @@ pub fn print_native_static_libs(sess: &Session, all_native_libs: &[NativeLibrary
     }
 }
 
+// Because windows-gnu target is meant to be self-contained for pure Rust code it bundles
+// own mingw-w64 libraries. These libraries are usually not compatible with mingw-w64
+// installed in the system. This breaks many cases where Rust is mixed with other languages
+// (e.g. *-sys crates).
+// We prefer system mingw-w64 libraries if they are available to avoid this issue.
+fn get_crt_libs_path(sess: &Session) -> Option<PathBuf> {
+    fn find_exe_in_path<P>(exe_name: P) -> Option<PathBuf>
+    where
+        P: AsRef<Path>,
+    {
+        for dir in env::split_paths(&env::var_os("PATH")?) {
+            let full_path = dir.join(&exe_name);
+            if full_path.is_file() {
+                return Some(fix_windows_verbatim_for_gcc(&full_path));
+            }
+        }
+        None
+    }
+
+    fn probe(sess: &Session) -> Option<PathBuf> {
+        if let (linker, LinkerFlavor::Gcc) = linker_and_flavor(&sess) {
+            let linker_path = if cfg!(windows) && linker.extension().is_none() {
+                linker.with_extension("exe")
+            } else {
+                linker
+            };
+            if let Some(linker_path) = find_exe_in_path(linker_path) {
+                let mingw_arch = match &sess.target.target.arch {
+                    x if x == "x86" => "i686",
+                    x => x,
+                };
+                let mingw_dir = format!("{}-w64-mingw32", mingw_arch);
+                // Here we have path/bin/gcc but we need path/
+                let mut path = linker_path;
+                path.pop();
+                path.pop();
+                // Based on Clang MinGW driver
+                let probe_path = path.join(&mingw_dir).join("lib");
+                if probe_path.exists() {
+                    return Some(probe_path);
+                };
+                let probe_path = path.join(&mingw_dir).join("sys-root/mingw/lib");
+                if probe_path.exists() {
+                    return Some(probe_path);
+                };
+            };
+        };
+        None
+    }
+
+    let mut system_library_path = sess.system_library_path.borrow_mut();
+    match &*system_library_path {
+        Some(Some(compiler_libs_path)) => Some(compiler_libs_path.clone()),
+        Some(None) => None,
+        None => {
+            let path = probe(sess);
+            *system_library_path = Some(path.clone());
+            path
+        }
+    }
+}
+
 pub fn get_file_path(sess: &Session, name: &str) -> PathBuf {
+    // prefer system {,dll}crt2.o libs, see get_crt_libs_path comment for more details
+    if sess.target.target.llvm_target.contains("windows-gnu") {
+        if let Some(compiler_libs_path) = get_crt_libs_path(sess) {
+            let file_path = compiler_libs_path.join(name);
+            if file_path.exists() {
+                return file_path;
+            }
+        }
+    }
     let fs = sess.target_filesearch(PathKind::Native);
     let file_path = fs.get_lib_path().join(name);
     if file_path.exists() {
@@ -1149,6 +1220,13 @@ fn link_args<'a, B: ArchiveBuilder<'a>>(
 
     // target descriptor
     let t = &sess.target.target;
+
+    // prefer system mingw-w64 libs, see get_crt_libs_path comment for more details
+    if cfg!(windows) && sess.target.target.llvm_target.contains("windows-gnu") {
+        if let Some(compiler_libs_path) = get_crt_libs_path(sess) {
+            cmd.include_path(&compiler_libs_path);
+        }
+    }
 
     cmd.include_path(&fix_windows_verbatim_for_gcc(&lib_path));
 

--- a/src/librustc_session/session.rs
+++ b/src/librustc_session/session.rs
@@ -133,6 +133,10 @@ pub struct Session {
     /// Mapping from ident span to path span for paths that don't exist as written, but that
     /// exist under `std`. For example, wrote `str::from_utf8` instead of `std::str::from_utf8`.
     pub confused_type_with_std_module: Lock<FxHashMap<Span, Span>>,
+
+    /// Path for libraries that will take preference over libraries shipped by Rust.
+    /// Used by windows-gnu targets to priortize system mingw-w64 libraries.
+    pub system_library_path: OneThread<RefCell<Option<Option<PathBuf>>>>,
 }
 
 pub struct PerfStats {
@@ -1069,6 +1073,7 @@ fn build_session_(
         driver_lint_caps,
         trait_methods_not_found: Lock::new(Default::default()),
         confused_type_with_std_module: Lock::new(Default::default()),
+        system_library_path: OneThread::new(RefCell::new(Default::default())),
     };
 
     validate_commandline_args_with_session_available(&sess);

